File: xox.py
import random as rnd
from collections import defaultdict
import operator

def takeinput(string,minvalue):
	while True:
		try:
			varname=int(input(f"{string}: "))
			if varname>minvalue and type(varname)==int:
				break
			print("What the f*** are you doing? Entry must exceed ",minvalue)
			if minvalue==0: print("Then why did you choose to play?")
		except Exception as e:
			print("Are you trolling? --> ",e)
		
	return varname

def isinarray(array,char):
	for player in array:
		if player[0]==char: return 1
		return 0
def createTable(m,n):
	mlist=[]
	for i in range(m):
		nlist=[]
		for j in range(n):
			nlist.append(" ")
		mlist.append(nlist)
	return mlist

def printTable(table,m,n):
	print()
	j=0
	while(j < 2*m+1):
		if j%2==0:
			for i in range(n):
				print(" -",end="",flush=True)
		else:
			for i in range(n):
				index=int((j-1)/2)
				print(f"|{table[index][i]}",end="",flush=True)
			print("|",end="",flush=True)
		print("\n",end="",flush=True)
		j+=1
	print()

def printElements(table,m,n):
	lentab=range(len(table))
	lenrow=range(len(table[0]))
	for row in table:
		if len(row)!=n: print()
	for j in lentab:
		length=len(table[j])
		if length!=n: print(f"{j}.row->{length} long")
		for i in lenrow:
			element=table[j][i]
			if element==" ": print(f"({i},{j})   space")
			else:print(f"({i},{j})   {element}")


def saveCoo(coolist,x,y,char):
	for player in coolist:
		if player[0]==char:
			player[1].append(x)
			player[2].append(y)
			return coolist
	
def findroundwinner(coolist,win_occ):
	for player in coolist:
		d=defaultdict(int)
		for i in player[1]:
			d[i] += 1
		result = max(list(d.items()), key=operator.itemgetter(1))
		
		if result[1]==win_occ: return player[0]
		d=defaultdict(int)
		for i in player[2]:
			d[i] += 1
		result = max(list(d.items()), key=operator.itemgetter(1))
		
		if result[1]==win_occ: return player[0]
	return 0
	
def xoxgame():

#game initials
	game=takeinput("How many rounds? (at least 1)",0)
	win_occ=takeinput("Win after how many occurences? (at least 3)",2)
	m=takeinput(f"Enter column (at least {win_occ})",win_occ-1)
	n=takeinput(f"Enter row (at least as {win_occ})",win_occ-1)
	playernum=takeinput("How many players? (at least 1)",0)
	
	coolist=[]
	scoreboard={}
#player names
	i=0
	while i < playernum:
		char=str(input(f"Alias of player (only one character) {i+1}: "))
		if len(char)!=1 or isinarray(coolist,char):
			print("Invalid alias with multiple characters or try different sequence!!! Do not oppose me!!!")
		else:
			 coolist.append([char,[],[]])
			 scoreboard[char]=0
			 i+=1
#######################	完成です!!!		 
	rnd.shuffle(coolist)
	print("\n################################################################################\n")
	print(f"{coolist[0][0]} starts!!!")
	
	lencoolist=len(coolist)
	 
	
	#number of rounds (<=game) and the only coefficient for scoring 
	countround=1
	
	#between rounds
	while countround<=game:
		playerturn=0
		#number of plays
		playcount=0
		table=createTable(m,n)
		#during a round
		while True:
			print(f"{countround}. round:\n")
			if playerturn==lencoolist: playerturn-=lencoolist
			char=coolist[playerturn][0]
		
			print(f"\n{char}'s turn")
			printTable(table,m,n)
#			print("ELEMENTS: ")
#			printElements(table,m,n)
			while True:
				y=int(input("Enter the x coordinate: "))
				x=int(input("Enter the y coordinate: "))
				
				if x<n and y<m:
					if table[x][y]==" ":
						table[x][y]=char
						break
					else: print("The square is not empty, is indeed occupied!!!")
						
				else: print("Invalid coordinates. Out of bounds!!!")
					
					
			coolist=saveCoo(coolist,x,y,char)
			scoreboard[char]+=countround
			playerturn+=1
			if playcount==playernum*(win_occ-1):
				roundwinner=findroundwinner(coolist,win_occ)
				if roundwinner!=0: 
					printTable(table,m,n)
	#				print("ELEMENTS: ")
	#				printElements(table,m,n)
					break
			playcount+=1
		#############################################################################################
		print("\n=================================================================================\n")
		
		print(f"{roundwinner} wins the round!!!\n")
		
		#SCOREBOARD AT THE END OF THE ROUND
		print("SCOREBOARD:")
		
		for player in scoreboard:
			print(player,": ",scoreboard[player])
		countround+=1
		
		
	
	#find the winner of the game
	gamewinner=max(scoreboard.items(), key=operator.itemgetter(1))[0]
	print("\n################################################################################\n")
	print(f"{gamewinner} wins the game with a score of {scoreboard[gamewinner]}!!!\n")
	#SCOREBOARD AT THE END OF THE GAME
	print("SCOREBOARD:")
	
	for player in scoreboard:
		print(player,": ",scoreboard[player])

if __name__ == "__main__":
	xoxgame()
	

File: xox.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "usefulfunc.h"

char** createTable(int m,int n)
{
	int i,j;
	char** table=(char**)malloc(m*sizeof(char*));
	for(i=0;i<m;++i)
	{
		table[i]=(char*)malloc(n*sizeof(char));
		for(j=0;j<n;++j)
		{
			table[i][j]=' ';
		}
	}
	
	return table;
}

void printTable(char** table,int m,int n)
{
	printf("\n");
	int j=0,i,index_int;
	float index;
	for(i=0;i<n;++i) printf(" %d",i);
	printf("\n");
	while(j < 2*m+1)
	{
		if(j%2==0)
		{
			for(i=0;i<n;++i) printf(" -");
		}
		else
		{
			for(i=0;i<n;++i)
			{
				index=(j-1)/2;
				index_int=(int)index;
				printf("|%c",table[index_int][i]);
			}
			printf("| %d",index_int);
		}
		printf("\n");
		++j;
	}
	printf("\n");
}

int findroundwinner(char** table,int m,int n,int win_occ,char ch)
{
	int count=1,i,j;
	//horizontal case
	for(i=0;i<m;++i)
		for(j=0;j<n-1;++j)
		{
			if(table[i][j]==ch && table[i][j+1]==ch) ++count;
			else count=1;
			if (count==win_occ) return 1;
		}
	count=1;
	//vertical case
	for(j=0;j<n;++j)
		for(i=0;i<m-1;++i)
		{
			if(table[i][j]==ch && table[i+1][j]==ch) ++count;
			else count=1;
			if (count==win_occ) return 1;
		}
	//cross case (left-right)
	count=1;
	i=n-win_occ;
	j=0;
	while(1)
	{
		if (table[j][i]==ch && table[j+1][i+1]==ch) ++count;
		else count=1;
		
		if (m<n)
		{
			if (i<=m-2 && j==m-2)
			{
				j-=(i-1);
				i=0;
			}
			else if(i==n-2 || j==m-2)
			{
				i-=(j+1);
				j=0;
			}
			else
			{
				++i;
				++j;
			}
		}
		else if(m>=n)
		{
			if (i==n-2 && j<n-2)
			{
				i-=(j+1);
				j=0;
			}
			else if (i==n-2 || j==m-2)
			{
				j-=(i-1);
				i=0;
			}
			else
			{
				++i;
				++j;
			}
		}
		if(count==win_occ) return 1;
		if (i==win_occ-2 && j==m-2) break;
	}
	//cross case (right-left)
	count=1;
	i=win_occ-1;
	j=0;
	while(1)
	{
		if(table[j][i]==ch && table[j+1][i-1]) ++count;
		else count=1;
		if(m<n)
		{
			if(i>=n-m+1 && j==m-2)
			{
				j-=(n-i-2);
				i=n-1;
			}
			else if(i==1 || j==m-2)
			{
				i+=(j+1);
				j=0;
			}
			else
			{
				--i;
				++j;
			}
		}
		else if(m>=n)
		{
			if(i==1 && j<n-2)
			{
				i+=(j+1);
				j=0;
			}
			else if(i==1 || j==m-2)
			{
				j-=(n-i-2);
				i=n-1;
			}
			else
			{
				i-=1;
				j+=1;
			}
		}
		if(count==win_occ) return 1;
		if(i==n-win_occ+1 && j==m-2) break;
	
	}
	return 0;
}

void shuffle(char *array, size_t n)
{
   time_t t;
   
   /* Intializes random number generator */
   srand((unsigned) time(&t));
   
    if (n > 1) 
    {
        size_t i;
        for (i = 0; i < n - 1; i++) 
        {
          size_t j = i + rand() / (RAND_MAX / (n - i) + 1);
          char t = array[j];
          array[j] = array[i];
          array[i] = t;
        }
    }
}

int takeinput(char* string,int var,int minvalue)
{
	int varname=0;
	while(1)
	{
		printf("%s%d): ",string,var);
		scanf("%d",&varname);
		getchar();
		if (varname>minvalue) break;
		printf("What the f*** are you doing? Entry must exceed %d\n",minvalue);
		if (!minvalue) printf("Then why did you choose to play?\n");
	
	}
	return varname;

}



int main()
{
	char** table;
	char character;
	int winner,scorebest,x,y;
	int game=takeinput("How many rounds? (at least ",1,0);
	int win_occ=takeinput("Win after how many occurences? (at least ",3,2);
	
	int m=takeinput("Enter column (at least ",win_occ,win_occ-1);
	int n=takeinput("Enter row (at least ",win_occ,win_occ-1);
	int playernum=takeinput("How many players? (at least ",1,0);
	
	//player names
	int i=0;
	char* ch=malloc(playernum*sizeof(char));
	int* scoreboard=calloc(playernum,sizeof(int));
	while(i < playernum)
	{
		printf("Alias of player (only one character) %d: ",i+1);
		scanf("%c",&ch[i]);
		getchar();
		++i;
	}

	//for (i=0;i<playernum;++i) printf("%c ",ch[i]);
	
	
	shuffle(ch,playernum);
	printf("\n################################################################################\n");
	printf("%c starts!!!",ch[0]);
	
	//number of rounds (<=game) and the only coefficient for scoring 
	int countround=1;
	//between rounds
	
	while (countround<=game)
	{
		int winner=0;
		int playerturn=0;
		
		//number of plays
		int playcount=0;
		
		table=createTable(m,n);
		//during a round
		while (1)
		{
			printf("%d. round:\n",countround);
			if (playerturn==playernum) playerturn-=playernum;
			character=ch[playerturn];
			printf("\n%c's turn\n",character);
			printTable(table,m,n);
			while (1)
			{
				
				printf("Enter the x coordinate: ");
				scanf("%d",&y);
				printf("Enter the y coordinate: ");
				scanf("%d",&x);
				
				if (x<m && y<n)
				{
					if (table[x][y]==' ')
					{
						table[x][y]=character;
						break;
					}
					else printf("The square is not empty, is indeed occupied!!!\n");
				}		
				else printf("Invalid coordinates. Out of bounds!!!\n");
			}
			scoreboard[playerturn]+=countround;
			++playerturn;
			if (playcount>=playernum*(win_occ-1))
			{
				for(i=0;i<playernum;++i)
				{
					if(findroundwinner(table,m,n,win_occ,ch[i]))
					{
						winner=ch[i];
						printTable(table,m,n);
						break;
					}
				}
				if(winner) break;
			}
			++playcount;
		}
		printf("\n=================================================================================\n");
		printf("%c wins the round!!!\n",winner);
		
		//SCOREBOARD AT THE END OF THE ROUND
		printf("SCOREBOARD:\n");
		
		for(i=0;i<playernum;++i)
		{
			printf("%c: %d\n",ch[i],scoreboard[i]);
		}
		
		++countround;
	}
	//find the winner of the game
	scorebest=nmax(scoreboard,playernum);
	winner=ch[scorebest];
	printf("\n################################################################################\n");
	
	printf("%c wins the game with a score of %d!!!\n",winner,scoreboard[scorebest]);
	
	//SCOREBOARD AT THE END OF THE GAME
	printf("SCOREBOARD:\n");
		
		for(i=0;i<playernum;++i)
		{
			printf("%c: %d\n",ch[i],scoreboard[i]);
		}
	
	
	for(i=0;i<m;++i) free(table[i]);
	free(table);
	
	free(ch);
	free(scoreboard);
	return 0;
}
